Pharo 9 TFFI updates

akgrant43 opened the pull request:
This PR makes two changes for Pharo 9 compatibility:

- Change parameter type from `ExternalAddress` to `void *` to prevent 'object not pinned' errors.
- Pharo 9 TFFI assumes functions with String return type are UTF8 and decodes

Changes:

- Change the parameter type in:

    SQLite3StatementExternalObject class>>finalizeResourceData:
    SQLite3DatabaseExternalObject class>>finalizeResourceData:

- Modify SQLite3Library>>utf8StringToPharo: to not decode the UTF8 in Pharo 9 and later.

akgrant43 commented:
The Glorp test failures are pre-existing, but not consistent.

For example, in Pharo 8 (without this PR applied) `GlorpDictionaryMappingTest>>#testDeleteSimpleTypeOneToMany` will pass when stepping through the Glorp code, but will often fail if run normally.  Enabling logging shows the following sequence of statements for PUBLISHER_TITLE (the primary table that is modified):

```
CREATE TABLE PUBLISHER_TITLE (PUBLISHER_ID int  NULL ,POSITION int  NULL ,TITLE text  NULL , CONSTRAINT PUBLISHER__TO_GR_PUBLISH_REF1 FOREIGN KEY (PUBLISHER_ID) REFERENCES GR_PUBLISHER (ID))


INSERT INTO PUBLISHER_TITLE (PUBLISHER_ID,POSITION,TITLE)  VALUES (?,?,?)  #(1 3 'Metadata-Based Persistence for Dummies')
INSERT INTO PUBLISHER_TITLE (PUBLISHER_ID,POSITION,TITLE)  VALUES (?,?,?)  #(2 1 'A book')
INSERT INTO PUBLISHER_TITLE (PUBLISHER_ID,POSITION,TITLE)  VALUES (?,?,?)  #(1 1 'Mastering ENVY/Developer')
INSERT INTO PUBLISHER_TITLE (PUBLISHER_ID,POSITION,TITLE)  VALUES (?,?,?)  #(1 2 'Principia Mathematica')

INSERT INTO PUBLISHER_TITLE (PUBLISHER_ID,POSITION,TITLE)  VALUES (?,?,?)  #(1 2 'Principia Mathematica')
DELETE FROM PUBLISHER_TITLE WHERE PUBLISHER_ID = ? AND POSITION = ? AND TITLE = ?  #(1 2 'Mastering ENVY/Developer')
DELETE FROM PUBLISHER_TITLE WHERE PUBLISHER_ID = ? AND POSITION = ? AND TITLE = ?  #(1 3 'Principia Mathematica')
```

This is clearly wrong as it is attempting to delete 'Principa Mathematica' with (publisher_id, position) pair (1 3), when the entry was created as (1 2).  Similarly, 'Mastering ENVY/Developer' is created with (1 1), but attempts to delete with (1 2)

Note that the SQL statements can differ between runs when failing.

gcotelli commented on src/SQLite3-Core/SQLite3Library.class.st:
```diff
@@ -868,7 +868,9 @@ SQLite3Library >> unix64ModuleName [
 	"Converts from SQLite UTF-8 to Pharo Multibyte Characters"
 	
-	^(ZnCharacterReadStream on: anUTF8String asByteArray readStream) upToEnd
+	^ SystemVersion current major >= 9
+		ifTrue: [ anUTF8String ]
+		ifFalse: [ (ZnCharacterReadStream on: anUTF8String asByteArray readStream) upToEnd ]
```
I think a cleaner solution is to move this method as an extension in other packages (one for Pharo-9 and the other for Pharo<9) and update the baseline to load the correct version of the package depending on the Pharo version.

akgrant43 replied:
Good suggestion, shall do...

gcotelli commented:
> The Glorp test failures are pre-existing, but not consistent.
> 
> For example, in Pharo 8 (without this PR applied) `GlorpDictionaryMappingTest>>#testDeleteSimpleTypeOneToMany` will pass when stepping through the Glorp code, but will often fail if run normally. Enabling logging shows the following sequence of statements for PUBLISHER_TITLE (the primary table that is modified):
> 
> ```
> CREATE TABLE PUBLISHER_TITLE (PUBLISHER_ID int  NULL ,POSITION int  NULL ,TITLE text  NULL , CONSTRAINT PUBLISHER__TO_GR_PUBLISH_REF1 FOREIGN KEY (PUBLISHER_ID) REFERENCES GR_PUBLISHER (ID))
> 
> 
> INSERT INTO PUBLISHER_TITLE (PUBLISHER_ID,POSITION,TITLE)  VALUES (?,?,?)  #(1 3 'Metadata-Based Persistence for Dummies')
> INSERT INTO PUBLISHER_TITLE (PUBLISHER_ID,POSITION,TITLE)  VALUES (?,?,?)  #(2 1 'A book')
> INSERT INTO PUBLISHER_TITLE (PUBLISHER_ID,POSITION,TITLE)  VALUES (?,?,?)  #(1 1 'Mastering ENVY/Developer')
> INSERT INTO PUBLISHER_TITLE (PUBLISHER_ID,POSITION,TITLE)  VALUES (?,?,?)  #(1 2 'Principia Mathematica')
> 
> INSERT INTO PUBLISHER_TITLE (PUBLISHER_ID,POSITION,TITLE)  VALUES (?,?,?)  #(1 2 'Principia Mathematica')
> DELETE FROM PUBLISHER_TITLE WHERE PUBLISHER_ID = ? AND POSITION = ? AND TITLE = ?  #(1 2 'Mastering ENVY/Developer')
> DELETE FROM PUBLISHER_TITLE WHERE PUBLISHER_ID = ? AND POSITION = ? AND TITLE = ?  #(1 3 'Principia Mathematica')
> ```
> 
> This is clearly wrong as it is attempting to delete 'Principa Mathematica' with (publisher_id, position) pair (1 3), when the entry was created as (1 2). Similarly, 'Mastering ENVY/Developer' is created with (1 1), but attempts to delete with (1 2)
> 
> Note that the SQL statements can differ between runs when failing.

Certainly, the changes don't look like something that will make this specific test fail.

akgrant43 commented:
Hmm, not sure why the Unit Tests are failing here.  The load and tests run successfully on my local machine.  Investigating...

gcotelli commented on src/BaselineOfSQLite3/BaselineOfSQLite3.class.st:
```diff
@@ -22,9 +23,14 @@ BaselineOfSQLite3 >> baseline: spec [
 				package: 'SQLite3-Glorp' with: [ spec requires: #('SQLite3-Core' 'Glorp-Core') ];
 				package: 'SQLite3-Glorp-Tests' with: [ spec requires: #('SQLite3-Glorp' 'Glorp-Tests') ].
+			"Pharo version compatibility"
+			versionPackage := SystemVersion current major >= 9 
+				ifTrue: [ 'SQLite3-Pharo9' ]
+				ifFalse: [ 'SQLite3-Pharo8' ].
+			spec package: versionPackage with: [ spec requires: #('SQLite3-Core') ].
```
I don't think this would work as you expected. Metacello is a bit special. Take a look at https://github.com/ba-st/Launchpad/blob/release-candidate/source/BaselineOfLaunchpad/BaselineOfLaunchpad.class.st 
I never found another way to make it load something conditional to the version than using the special `#pharo{V}.x` spec.

gcotelli replied:
This example is a package for Pharo 7 but hopefully, you should get an idea of what to do

akgrant43 replied:
Thanks Gabriel!  I was wanting to avoid duplicating the bulk of the baseline, but didn't know about `#group:with:`.  I'll rewrite the method if I can't get the above to work on the CI.

But, running it on my PC... :-)

```
$ bin/smalltalkci --headful /home/alistair/pharo/Pharo-SQLite3/.smalltalk.ston 
1) Squeak64-trunk	10) Squeak32-4.6	19) Pharo32-stable	28) GemStone64-3.5.0	37) Moose64-8.0
2) Squeak64-5.3		11) Squeak32-4.5	20) Pharo32-alpha	29) GemStone64-3.4.3	38) Moose64-7.0
3) Squeak64-5.2		12) Pharo64-stable	21) Pharo32-9.0		30) GemStone64-3.3.9	39) Moose32-trunk
4) Squeak64-5.1		13) Pharo64-alpha	22) Pharo32-8.0		31) GemStone64-3.3.2	40) Moose32-8.0
5) Squeak32-trunk	14) Pharo64-9.0		23) Pharo32-7.0		32) GemStone64-3.3.0	41) Moose32-7.0
6) Squeak32-5.3		15) Pharo64-8.0		24) Pharo32-6.0		33) GemStone64-3.2.12	42) Moose32-6.1
7) Squeak32-5.2		16) Pharo64-7.0		25) Pharo32-5.0		34) GemStone64-3.1.0.6	43) Moose32-6.0
8) Squeak32-5.1		17) Pharo64-6.1		26) Pharo32-4.0		35) Moose64-trunk
9) Squeak32-5.0		18) Pharo64-6.0		27) Pharo32-3.0		36) Moose64-9.0
Choose Smalltalk image: 14
Preparing folders...
Starting Pharo build...
Downloading Pharo64-9.0 image...
Downloading the latest 90 Image:
    http://files.pharo.org/get-files/90/pharoImage-x86_64.zip
Pharo.image
 > Time to run: 3.936s
Preparing Pharo image...
Downloading Pharo64-9.0 vm...
Downloading the latest pharoVM:
	http://files.pharo.org/get-files/90/pharo-vm-Linux-x86_64-stable.zip
pharo-vm/pharo
Creating starter scripts pharo and pharo-ui
 > Time to run: 7.074s
Image Information
Image
-----
/home/alistair/pharo/smalltalkCI/_builds/2021_09_21_20_34_22/TravisCI.image
Pharo9.0.0
Build information: Pharo-9.0.0+build.1549.sha.eae2e955c11d5c166543f2fc903c4bccf3e2b8f4 (64 Bit)
Unnamed


 > Done in 2.000ms.
Loading project...

 > Done in 1392.000ms.
Testing project...
Running suite "Pharo64-9.0" with 3 tests.
.....................................................................................Finished running suite: Pharo64-9.0

 > Done in 201.000ms.

##################################################
# Pharo64-9.0                                    #
# 85 Tests with 0 Failures and 0 Errors in 0.20s #
##################################################

(85 tests passed)

SQLiteBaseConnectionTest
 ✓ #testBadOpen (3ms)
 ✓ #testBasicExecuteOk (2ms)
 ✓ #testBasicExecuteSyntaxError (0ms)
 ✓ #testBooleanColumnTypeMismatch (0ms)
 ✓ #testBooleanObject (1ms)
 ✓ #testBooleanWriteIntegerOneThenRead (1ms)
 ✓ #testBooleanWriteIntegerTwoThenRead (0ms)
 ✓ #testBooleanWriteIntegerZeroThenRead (0ms)
 ✓ #testBooleanWriteNilThenRead (0ms)
 ✓ #testBooleanWriteSQLNullThenRead (0ms)
 ✓ #testBooleanWriteStringFalseThenRead (0ms)
 ✓ #testBooleanWriteStringTrueThenRead (0ms)
 ✓ #testBooleanWriteStringTwoThenRead (0ms)
 ✓ #testBooleanWriteThenRead (0ms)
 ✓ #testColumnNames (0ms)
 ✓ #testColumnNamesBobbyStop (0ms)
 ✓ #testColumnNamesNoSuchTable (0ms)
 ✓ #testConstraintViolation (0ms)
 ✓ #testDataValuesAvailable (0ms)
 ✓ #testDateAndTime (1ms)
 ✓ #testDeclaredColumnTypes (0ms)
 ✓ #testDefaultMultilingualStrings (0ms)
 ✓ #testDoubleClose (0ms)
 ✓ #testFloat (0ms)
 ✓ #testInteger (0ms)
 ✓ #testMultilingualStrings (0ms)
 ✓ #testMultilingualTableName (0ms)
 ✓ #testObjectBlob (1ms)
 ✓ #testPrepareStep (1ms)
 ✓ #testPrepareStepSelective (0ms)
 ✓ #testResultSetColumnIntrospection (0ms)
 ✓ #testScaledDecimal (0ms)
 ✓ #testTableNames (0ms)
 ✓ #testTableNamesOfEmptyDatabase (0ms)
 ✓ #testTime (0ms)

SQLiteConnectionTest
 ✓ #testBackup (104ms)
 ✓ #testChanges (2ms)
 ✓ #testExactlyOneRow (6ms)
 ✓ #testExactlyOneRowGotNone (5ms)
 ✓ #testExecuteDelete (0ms)
 ✓ #testExecuteDrop (0ms)
 ✓ #testExecuteNoResult (0ms)
 ✓ #testExecuteValueInsertArray (0ms)
 ✓ #testExecuteValueInsertBooleanNilThenSelect (0ms)
 ✓ #testExecuteValueInsertBooleanThenSelect (0ms)
 ✓ #testExecuteValueInsertChanges (0ms)
 ✓ #testExecuteValueInsertThenSelect (0ms)
 ✓ #testExecuteValueInsertThenSelect2 (0ms)
 ✓ #testExecuteValueInsertThenSelectDateTime (0ms)
 ✓ #testExecuteWithInsertArray (0ms)
 ✓ #testExecuteWithInsertBooleanNilThenSelect (0ms)
 ✓ #testExecuteWithInsertBooleanSQLNullThenSelect (0ms)
 ✓ #testExecuteWithInsertBooleanThenSelect (0ms)
 ✓ #testExecuteWithInsertChanges (0ms)
 ✓ #testExecuteWithInsertDictionary (0ms)
 ✓ #testExecuteWithInsertMultilingualStringsThenSelect (0ms)
 ✓ #testExecuteWithInsertNoDataThenSelect (0ms)
 ✓ #testExecuteWithInsertThenSelect (0ms)
 ✓ #testExecuteWithInsertThenSelect2 (0ms)
 ✓ #testExecuteWithInsertThenSelect3 (0ms)
 ✓ #testExecuteWithInsertThenSelectChineseTableName (0ms)
 ✓ #testExecuteWithInsertThenSelectDateTime (0ms)
 ✓ #testExecuteWithInsertThenSelectJapaneseTableName (0ms)
 ✓ #testExecuteWithToFail (0ms)
 ✓ #testExecuteWithToFail2 (0ms)
 ✓ #testGetAutocommit (0ms)
 ✓ #testGlorpAutoID (0ms)
 ✓ #testGlorpCompositeKey (0ms)
 ✓ #testGlorpDoubleInsertPrimaryKey (0ms)
 ✓ #testGlorpUpdate (0ms)
 ✓ #testGlorpUpdateTwice (0ms)
 ✓ #testInsertThenSelectInTransaction (0ms)
 ✓ #testRows (29ms)
 ✓ #testStatementReadOnly (1ms)
 ✓ #testValueInsertThenSelectInTransaction (0ms)

SQLite3RowTest
 ✓ #testAsArray (0ms)
 ✓ #testAsCombinedDictionary (0ms)
 ✓ #testAsDictionary (1ms)
 ✓ #testColumnAccessingIfAbsent (0ms)
 ✓ #testColumnAccessingOnEmptyRow (0ms)
 ✓ #testColumnAccessingOnFilledRow (0ms)
 ✓ #testColumnNamesOnEmptyRow (0ms)
 ✓ #testColumnNamesOnFilledRow (0ms)
 ✓ #testColumnPuttingOnEmptyRow (0ms)
 ✓ #testColumnPuttingOnFilledRow (0ms)


  Executed 85 Tests with 0 Failures and 0 Errors in 0.20s.

Skipping coveralls upload.
```

I'll keep looking.

akgrant43 commented:
Help! :-)

This looks like a CI issue to me.  I can run smalltalkCI locally and all tests pass for both Pharo 8 and 9.  I've refactored the baseline as Gabriel kindly suggested, and smalltalkCI still passes locally.

The errors kind of look like it is using the old baseline with the new repository code, like the image was saved with the baseline in it.

Any suggestions?

gcotelli commented:
I will check the code and let you know if I find anything. Taking a quick look seems to be ok so I don't understand why it fails.

akgrant43 commented:
> I will check the code and let you know if I find anything. Taking a quick look seems to be ok so I don't understand why it fails.

Thanks!

akgrant43 commented:
Hi Garbriel,
Thanks very much for fixing and merging the PR!
What was the significant change that made the baseline load correctly?  From what I can see, it looks like the only thing you changed was specifying `#group:with:` immediately after `#package:` (the logging obviously shouldn't have any effect, and grouping pharo 7 & 8 together in `#versionSpecificBaseline:` shouldn't matter).

gcotelli commented:
I've removed `blessing: #baseline` and changed the dependency on `SQLite3-Core-Tests` package to be against the `Core` group instead of a package. I think this last change was the one that make the project load. 

The rest are cosmetic changes.

akgrant43 commented:
> I've removed `blessing: #baseline` and changed the dependency on `SQLite3-Core-Tests` package to be against the `Core` group instead of a package. I think this last change was the one that make the project load.
> 
> The rest are cosmetic changes.

Weird.  But thanks for the explanation!
